common/client/multi_node_test.go

```diff
@@ -0,0 +1,630 @@
+package client
+
+import (
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/utils/tests"
+
+	"github.com/smartcontractkit/chainlink/v2/common/chains/client"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/config"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type multiNodeRPCClient RPC[types.ID, *utils.Big, Hashable, Hashable, any, Hashable, any, any,
+	types.Receipt[Hashable, Hashable], Hashable, types.Head[Hashable]]
+
+type testMultiNode struct {
+	*multiNode[types.ID, *utils.Big, Hashable, Hashable, any, Hashable, any, any,
+		types.Receipt[Hashable, Hashable], Hashable, types.Head[Hashable], multiNodeRPCClient]
+}
+
+type multiNodeOpts struct {
+	logger              logger.Logger
+	selectionMode       string
+	leaseDuration       time.Duration
+	noNewHeadsThreshold time.Duration
+	nodes               []Node[types.ID, types.Head[Hashable], multiNodeRPCClient]
+	sendonlys           []SendOnlyNode[types.ID, multiNodeRPCClient]
+	chainID             types.ID
+	chainType           config.ChainType
+	chainFamily         string
+	sendOnlyErrorParser func(err error) client.SendTxReturnCode
+}
+
+func newTestMultiNode(t *testing.T, opts multiNodeOpts) testMultiNode {
+	if opts.logger == nil {
+		opts.logger = logger.TestLogger(t)
+	}
+
+	result := NewMultiNode[types.ID, *utils.Big, Hashable, Hashable, any, Hashable, any, any,
+		types.Receipt[Hashable, Hashable], Hashable, types.Head[Hashable], multiNodeRPCClient](opts.logger,
```

Comment:
Not a big fan of Hashable being used as a Fee, but I guess we can keep it as it is now since there are no hard constraints.